### PR TITLE
Re-implementation of sampler type

### PR DESCRIPTION
**Pull request by AlexeySotkin**
Emit sampler as opaque type.

### CODE REVIEW

**Comment by AlexeySotkin** [comment]
@yxsamliu , could you take a look.

File: lib/CodeGen/CodeGenModule.cpp [hunk 1]
```diff
@@ -3718,3 +3718,50 @@ void CodeGenModule::EmitOMPThreadPrivateDecl(const OMPThreadPrivateDecl *D) {
...
+  if (!ConstSamplerTy ) {
+    llvm::Type* Elements[] = {Int32Ty, Int32Ty, Int32Ty};
+    ConstSamplerTy = llvm::StructType::create(VMContext, Elements,
+                                              "spirv.ConstantSampler");
+  }
+  const llvm::ConstantInt *CI = static_cast<llvm::ConstantInt*>(C);
+  const uint64_t SamplerValue = CI->getValue().getZExtValue();
+  // 32-bit value of sampler's initializer is interpreted as
+  // bit-field with the following structure:
+  // |unspecified|Filter|Addressing Mode| Normalized Coords|
```
Comment (yxsamliu) [inline lib/CodeGen/CodeGenModule.cpp:3743]: Filter should be Filter+1

File: test/CodeGenOpenCL/address-spaces-ocl20.cl [hunk 2]
```diff
@@ -13,7 +13,8 @@ global int *global baz_ptr = &baz;
 // CHECK: @baz_ptr = addrspace(1) global i32 addrspace(1)* @baz, align 4
 
 const sampler_t sampler = 10;
```
Comment (yxsamliu) [inline test/CodeGenOpenCL/address-spaces-ocl20.cl]: invalid value for sampler

File: lib/CodeGen/CGExprScalar.cpp [hunk 3]
```diff
@@ -1570,10 +1570,21 @@ Value *ScalarExprEmitter::VisitCastExpr(CastExpr *CE) {
 
   case CK_IntToOCLSampler: {
     assert(DestTy->isSamplerT() && "CK_IntToOCLSampler cast to non sampler type");
-    return Visit(E);
+    if (!CGF.CGM.getLangOpts().CLSamplerOpaque)
+      return Visit(E);
+    if (const CastExpr* SCE = dyn_cast<CastExpr>(E)) {
```
Comment (yxsamliu) [inline lib/CodeGen/CGExprScalar.cpp:1575]: why not using cast? same with the following 2 dyn_cast

File: lib/CodeGen/CGExprScalar.cpp [hunk 4]
```diff
@@ -1570,10 +1570,21 @@ Value *ScalarExprEmitter::VisitCastExpr(CastExpr *CE) {
...
+      if (const DeclRefExpr *DRE = dyn_cast<DeclRefExpr>(SCE->getSubExpr())) {
+        if (const VarDecl *VD = dyn_cast<VarDecl>(DRE->getDecl())) {
+          if (const Expr* Init = VD->getInit())
```
Comment (yxsamliu) [inline lib/CodeGen/CGExprScalar.cpp]: shouldn't this be always non-zero? why not add assert

File: lib/CodeGen/CGDecl.cpp [hunk 5]
```diff
@@ -273,7 +273,15 @@ static bool hasNontrivialDestruction(QualType T) {
 llvm::GlobalVariable *
 CodeGenFunction::AddInitializerToStaticVarDecl(const VarDecl &D,
                                                llvm::GlobalVariable *GV) {
-  llvm::Constant *Init = CGM.EmitConstantInit(D, this);
+  llvm::Constant *Init = nullptr;
+  if(D.getType()->isSamplerT() && CGM.getLangOpts().CLSamplerOpaque) {
+    if (const CastExpr * CE = dyn_cast_or_null<CastExpr>(D.getInit())) {
```
Comment (yxsamliu) [inline lib/CodeGen/CGDecl.cpp]: why do we need this? shouldn't all int to sampler casts be handled by CGExprConstant.cpp and CGExprScalar.cpp?

File: lib/CodeGen/CGCall.cpp [hunk 6]
```diff
@@ -2819,6 +2818,25 @@ void CodeGenFunction::EmitCallArg(CallArgList &args, const Expr *E,
     return;
   }
 
+  if (type->isSamplerT() && CGM.getLangOpts().CLSamplerOpaque) {
```
Comment (yxsamliu) [inline lib/CodeGen/CGCall.cpp]: why do we need this? shouldn't all int to sampler casts be handled by CGExprConstant.cpp and CGExprScalar.cpp?

File: lib/CodeGen/CodeGenModule.cpp (hunk 1 above)
Comment (AlexeySotkin) [inline lib/CodeGen/CodeGenModule.cpp:3743]: Why? I think Filter should be ok.

File: test/CodeGenOpenCL/address-spaces-ocl20.cl (hunk 2 above)
Comment (AlexeySotkin) [inline test/CodeGenOpenCL/address-spaces-ocl20.cl]: Fixed

File: lib/CodeGen/CGCall.cpp (hunk 6 above)
Comment (AlexeySotkin) [inline lib/CodeGen/CGCall.cpp]: Removed

File: lib/CodeGen/CGDecl.cpp (hunk 5 above)
Comment (AlexeySotkin) [inline lib/CodeGen/CGDecl.cpp]: Removed

File: lib/CodeGen/CGExprScalar.cpp (hunk 4 above)
Comment (AlexeySotkin) [inline lib/CodeGen/CGExprScalar.cpp]: Fixed

File: lib/CodeGen/CGExprScalar.cpp (hunk 3 above)
Comment (AlexeySotkin) [inline lib/CodeGen/CGExprScalar.cpp:1575]: In some cases E is not a CastExpr, so cast will fail with assert. The same for the following 2 dyn_cast.

File: lib/CodeGen/CodeGenModule.cpp (hunk 1 above)
Comment (yxsamliu) [inline lib/CodeGen/CodeGenModule.cpp:3743]: The 4-5 bits of a sampler value used in opencl source code has enum value:
define CLK_FILTER_NEAREST              0x10
define CLK_FILTER_LINEAR                  0x20
corresponding to 1 and 2 in bits 4-5.

SPIR-V defines filter enum as: nearest=0, linear=1, so the actual bits are greater than the SPIR-V value by 1. If we denote the value used by SPIR-V as `Filter`, the bits should be `Filter+1`.

**Comment by yxsamliu** [comment]
Can you squash 3 commits into 1 commit? Thanks.

File: lib/CodeGen/CodeGenModule.cpp (hunk 1 above)
Comment (AlexeySotkin) [inline lib/CodeGen/CodeGenModule.cpp:3743]: Right, operation you described is used in SPIRVReader, during translation of SPIRV to LLVM. Here we convert these bits of sampler value to values of the SPIRV enum, i.e. 1,2=>0,1. That is why I decrement Filter.

Comment (yxsamliu) [inline lib/CodeGen/CodeGenModule.cpp:3743]: How about add a comment to clarify about the SPIR-V Filter value and bits 4-5. It is not so obvious.

Comment (AlexeySotkin) [inline lib/CodeGen/CodeGenModule.cpp:3743]: OK, I'll add some clarifications

**Comment by AlexeySotkin** [comment]
I have added clarifications and squashed the commits.
BTW did you look at KhronosGroup/SPIRV-LLVM/pull/127 ?

File: lib/CodeGen/CGOpenCLRuntime.cpp [hunk 7]
```diff
@@ -77,9 +77,9 @@ llvm::Type *CGOpenCLRuntime::convertOpenCLSpecificType(const Type *T) {
     return llvm::PointerType::get(llvm::StructType::create(
                            Ctx, "opencl.image3d_t"), ImgAddrSpc);
   case BuiltinType::OCLSampler:
-    if (CGM.getLangOpts().CLKeepSamplerType)
-      return llvm::StructType::create(Ctx, llvm::IntegerType::get(Ctx, 32),
-                                      "opencl.sampler_t");
+    if (CGM.getLangOpts().CLSamplerOpaque)
+      return llvm::PointerType::get(llvm::StructType::create(
+                           Ctx, "spirv.Sampler"), 0);
```
Comment (yxsamliu) [inline lib/CodeGen/CGOpenCLRuntime.cpp]: I think we'd better put sampler in constant addr space.

Comment (AlexeySotkin) [inline lib/CodeGen/CGOpenCLRuntime.cpp]: What's the point?

Comment (yxsamliu) [inline lib/CodeGen/CGOpenCLRuntime.cpp]: This helps backends for optimization since they know the samplers do not change.

Comment (AlexeySotkin) [inline lib/CodeGen/CGOpenCLRuntime.cpp]: Done

**Pull request merged**